### PR TITLE
Refactor Loki generators to use new functional pattern

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

### CODE REVIEW

**Comment by kakkoyun** [comment]
@metalmatze @periklis This ready for review. I tried to keep changes to generated files minimal. Also, I tried to stick to the existing patterns. I'm open to suggestions tell me what you think.

File: components/loki.libsonnet [hunk 1]
```diff
@@ -1,28 +1,62 @@
...
+  version: error 'must provide version',
+  image: error 'must provide image',
+  replicas: error 'must provide replicas',
+  objectStorageConfig: error 'must provide object storage config',
+  queryConcurrency: 32,
+  queryParallelism: 32,  // Defaults to queryConcurrency because single query-frontend replica
+  ports: {
+    gossip: 7946,
+  },
+  replicationFactor: 0,
```
Comment (periklis) [inline components/loki.libsonnet:18]: The default should be always 1.

File: components/loki.libsonnet [hunk 2]
```diff
@@ -1,28 +1,62 @@
...
+  namespace: error 'must provide namespace',
+  version: error 'must provide version',
+  image: error 'must provide image',
+  replicas: error 'must provide replicas',
+  objectStorageConfig: error 'must provide object storage config',
+  queryConcurrency: 32,
+  queryParallelism: 32,  // Defaults to queryConcurrency because single query-frontend replica
+  ports: {
+    gossip: 7946,
+  },
```
Comment (periklis) [inline components/loki.libsonnet:17]: What is the reason to expose this port per config?

File: components/observatorium.libsonnet [hunk 3]
```diff
@@ -118,9 +138,8 @@ local api = (import 'observatorium/observatorium-api.libsonnet');
     } + {
       ['thanos-' + name]: obs.thanos.manifests[name]
       for name in std.objectFields(obs.thanos.manifests)
-    } + if std.length(obs.config.loki) != 0 then {
+    } + {
       ['loki-' + name]: obs.loki.manifests[name]
       for name in std.objectFields(obs.loki.manifests)
-    }
-    else {},
+    },
```
Comment (periklis) [inline components/observatorium.libsonnet:144]: Is the `if std.length(obs.config.loki) != 0` still ensured? IIRC, this check was added by ACM folks to deploy observatorium w/o loki.

**Review by periklis — DISMISSED** [review]
Looks in general good to me.

File: components/observatorium.libsonnet (hunk 3 above)
Comment (kakkoyun) [reply components/observatorium.libsonnet:144]: We now encourage an alternative way to build up the configuration in the downstream repo. Rather than inheritance, we want them to compose theirs on setup utilizing functions. So this should be fined.

File: components/loki.libsonnet (hunk 2 above)
Comment (kakkoyun) [reply components/loki.libsonnet:17]: I don't actually get the question, can you maybe rephrase? This is not exposed. This just a top-level configuration that we utilize whenever we need to access the ports.

Comment (periklis) [reply components/loki.libsonnet:17]: My intention was to say, that folks using the new pattern should be able to override this port.

**Review by periklis — COMMENTED** [review]
Awesome work! :rocket:

File: components/loki.libsonnet (hunk 2 above)
Comment (kakkoyun) [reply components/loki.libsonnet:17]: 👍 Yes, it's totally possible and cleaner.

File: components/loki.libsonnet [hunk 4]
```diff
@@ -1,29 +1,329 @@
-{
-  local loki = self,
+// These are the defaults for this components configuration.
+// When calling the function to generate the component's manifest,
+// you can pass an object structured like the default to overwrite default values.
+local defaults = {
+  local defaults = self,
+
+  name: error 'must provide name',
```
Comment (metalmatze) [inline components/loki.libsonnet:7]: It should be fine to set a default name here.

**Review by metalmatze — COMMENTED** [review]
Overall going in a good direction. I think there can be still a clearer separation between the individual components of Loki and thus splitting the defaults object quite drastically. Nothing that needs to be done in this PR though.
